File: gcode/src/main/java/org/qw3rtrun/p3d/g/decoder/TemperatureReportedDecoder.java
package org.qw3rtrun.p3d.g.decoder;

import org.qw3rtrun.p3d.g.event.OkTemperatureReported;
import org.qw3rtrun.p3d.g.event.TemperatureReport;
import org.qw3rtrun.p3d.g.event.TemperatureReported;
import org.qw3rtrun.p3d.g.event.TemperatureReportedEvent;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class TemperatureReportedDecoder implements GEventDecoder<TemperatureReportedEvent> {

    private static final Pattern TEMP_REPORT_PATTERN = Pattern.compile(
            "(?>([TtBb]\\d?): *[+-]?((?>[0-9]*.)?[0-9]+) */ *((?>[0-9]*.)?[0-9]+) *)|(?>([TtBb]?@\\d?): *[+-]?([0-9]+) *)"
    );

    @Override
    public Optional<TemperatureReportedEvent> decode(String line) {
        Matcher matcher = TEMP_REPORT_PATTERN.matcher(line);
        Map<String, Double> temps = new HashMap<>();
        Map<String, Double> targets = new HashMap<>();
        Map<String, Integer> pows = new HashMap<>();
        while (matcher.find()) {
            processGroup(matcher, temps, targets, pows);
        }
        var temp = temps.getOrDefault("T", temps.get("t"));
        var power = pows.getOrDefault("@", pows.get("t"));
        var target = targets.getOrDefault("T", targets.get("t"));
        var bedTemp = temps.getOrDefault("B", temps.get("b"));
        var bedPower = pows.getOrDefault("B@", pows.get("b"));
        var bedTarget = targets.getOrDefault("B", targets.get("b"));
        if (temp == null || bedTemp == null || target == null || bedTarget == null || power == null || bedPower == null) {
            return Optional.empty();
        }
        var treport = new TemperatureReport(temp, target, power);
        var breport = new TemperatureReport(bedTemp, bedTarget, bedPower);
        if ("ok".equalsIgnoreCase(line.substring(0, 2))) {
            return Optional.of(new OkTemperatureReported(treport, breport));
        }
        return Optional.of(new TemperatureReported(treport, breport));
    }

    public void processGroup(Matcher matcher, Map<String, Double> temps, Map<String, Double> targets, Map<String, Integer> powers) {
        matcher.group();
        if (matcher.group(3) != null) {
            processTempReport(matcher, temps, targets);
        } else {
            processPower(matcher, powers);
        }
    }

    private void processPower(Matcher matcher, Map<String, Integer> powers) {
        powers.put(matcher.group(4), Integer.valueOf(matcher.group(5)));
    }

    private void processTempReport(Matcher matcher, Map<String, Double> temps, Map<String, Double> targets) {
        temps.put(matcher.group(1), Double.valueOf(matcher.group(2)));
        targets.put(matcher.group(1), Double.valueOf(matcher.group(3)));
    }
}

File: gcode/src/main/kotlin/org/qw3rtrun/p3d/g/marlin/decoder/TemperatureReportedDecoder.kt
package org.qw3rtrun.p3d.g.marlin.decoder

import org.qw3rtrun.p3d.core.msg.TemperatureReport
import org.qw3rtrun.p3d.core.msg.TemperatureReportedEvent
import org.qw3rtrun.p3d.g.marlin.event.OkTemperatureReported
import org.qw3rtrun.p3d.g.marlin.event.TemperatureReported
import java.util.Optional
import java.util.regex.Matcher
import java.util.regex.Pattern

class TemperatureReportedDecoder : GEventDecoder<TemperatureReportedEvent> {

    override fun decode(line: String): Optional<TemperatureReportedEvent> {
        val matcher = TEMP_REPORT_PATTERN.matcher(line)
        val temps = mutableMapOf<String, Double>()
        val targets = mutableMapOf<String, Double>()
        val pows = mutableMapOf<String, Int>()
        while (matcher.find()) {
            processGroup(matcher, temps, targets, pows)
        }
        val temp = temps["T"] ?: temps["t"]
        val power = pows["@"] ?: pows["t"]
        val target = targets["T"] ?: targets["t"]
        val bedTemp = temps["B"] ?: temps["b"]
        val bedPower = pows["B@"] ?: pows["b"]
        val bedTarget = targets["B"] ?: targets["b"]
        if (temp == null || bedTemp == null || target == null || bedTarget == null || power == null || bedPower == null) {
            return Optional.empty()
        }
        val treport = TemperatureReport(temp, target, power)
        val breport = TemperatureReport(bedTemp, bedTarget, bedPower)
        if (line.length >= 2 && "ok".equals(line.substring(0, 2), ignoreCase = true)) {
            return Optional.of(OkTemperatureReported(treport, breport))
        }
        return Optional.of(TemperatureReported(treport, breport))
    }

    fun processGroup(
        matcher: Matcher,
        temps: MutableMap<String, Double>,
        targets: MutableMap<String, Double>,
        powers: MutableMap<String, Int>
    ) {
        if (matcher.group(3) != null) {
            processTempReport(matcher, temps, targets)
        } else {
            processPower(matcher, powers)
        }
    }

    private fun processPower(matcher: Matcher, powers: MutableMap<String, Int>) {
        // spec 9: a digit run too wide for Int is malformed input off the wire, not a crash. Drop
        // the field; the completeness check in decode() then turns the whole line into absence.
        // The group is [0-9]+, so toIntOrNull() never sees a non-ASCII digit here.
        val power = matcher.group(5).toIntOrNull() ?: return
        powers[matcher.group(4)] = power
    }

    private fun processTempReport(
        matcher: Matcher,
        temps: MutableMap<String, Double>,
        targets: MutableMap<String, Double>
    ) {
        // spec 9: same rule as processPower - an unparsable value drops its field rather than
        // throwing. Both halves are stored or neither, so a half-read field cannot be decoded.
        val temp = matcher.group(2).toDoubleOrNull() ?: return
        val target = matcher.group(3).toDoubleOrNull() ?: return
        temps[matcher.group(1)] = temp
        targets[matcher.group(1)] = target
    }

    companion object {
        private val TEMP_REPORT_PATTERN = Pattern.compile(
            // The dot is escaped: an unescaped `.` in `[0-9]*.` matched any character, so
            // `B:x1` parsed as a number and reached toDouble(). spec 3.1 shape: digits, an
            // optional `.` and more digits.
            "(?>([TtBb]\\d?): *[+-]?((?>[0-9]*\\.)?[0-9]+) */ *((?>[0-9]*\\.)?[0-9]+) *)|(?>([TtBb]?@\\d?): *[+-]?([0-9]+) *)"
        )
    }
}
